src/storage/gc_worker.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::engine::{Engine, ScanMode};
+use super::mvcc::{MvccReader, MvccTxn, MAX_TXN_WRITE_SIZE};
+use super::txn::GC_BATCH_SIZE;
+use super::{Callback, Error, Key, Result};
+use kvproto::kvrpcpb::Context;
+use std::fmt::{self, Display, Formatter};
+use std::sync::{Arc, Mutex};
+use util::worker::{self, Builder, Runnable, Worker};
+
+struct GCTask {
+    pub ctx: Context,
+    pub safe_point: u64,
+    pub callback: Callback<()>,
+}
+
+impl Display for GCTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let epoch = match self.ctx.region_epoch.as_ref() {
+            None => "None".to_string(),
+            Some(e) => format!("{{{}, {}}}", e.conf_ver, e.version),
+        };
+        write!(
+            f,
+            "GC task at region {}, epoch {}, safe_point {}",
+            self.ctx.region_id, epoch, self.safe_point
+        )
+    }
+}
+
+// `GCRunner` is used to perform GC on the engine
+struct GCRunner {
+    engine: Box<Engine>,
+    gc_ratio_threshold: f64,
+}
+
+impl GCRunner {
+    pub fn new(engine: Box<Engine>, gc_ratio_threshold: f64) -> GCRunner {
+        GCRunner {
+            engine,
+            gc_ratio_threshold,
+        }
+    }
+
+    fn gc_scan_keys(
+        &self,
+        ctx: &Context,
+        safe_point: u64,
+        next_scan_key: Option<Key>,
+    ) -> Result<(Option<Vec<Key>>, Option<Key>)> {
+        let snapshot = self.engine.snapshot(ctx)?;
+        let mut reader = MvccReader::new(
+            snapshot,
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            None,
+            None,
+            ctx.get_isolation_level(),
+        );
+        if !reader.need_gc(safe_point, self.gc_ratio_threshold) {
+            Ok((None, None))
+        } else {
+            reader
+                .scan_keys(next_scan_key, GC_BATCH_SIZE)
```

Comment:
Can we move `GC_BATCH_SIZE` into this mod?

Reply:
I think it's better to leave it in `txn` mod, for that there are some other `XXX_BATCH_SIZE` declared in it.

Reply:
How about move it (and other `XXX_BATCH_SIZE`s) to `txn/mod.rs`?